routers/events.router.js

```diff
@@ -1,32 +1,173 @@
+    if (maxPrice) {
+      filterQuery.price = { $lte: maxPrice };
+    }
+
+    if (type) {
+      filterQuery.type = type;
+    }
+
+    if (requiresApproval !== undefined) {
+      filterQuery.approvalRequired = requiresApproval === `true` ? true : false;
+    }
+
+    if (isGroupEvent !== undefined) {
+      if (isGroupEvent === `true`) {
+        filterQuery.$or = [
+          { "attendees.maximum": { $gt: 1 } },
+          { "attendees.maximum": { $exists: false } }
+        ]
+      } else {
+        filterQuery["attendees.maximum"] = 1;
+      }
+    }
+
+    if (search) {
+      filterQuery.$or = [
+        { title: new RegExp(search, `ig`) },
+        { description: new RegExp(search, `ig`) }
+      ]
+    }
+
+
+    console.log({ filterQuery })
+    let filteredEvents = await Event.find(
+      filterQuery,
+      {},
+      { sort: { startAt: 1 }, skip: page * 2, limit: 2 }
+    )
+      .populate({
+        path: `creator`,
+        select: { password: 0, email: 0, __v: 0 }
+      });
+
+    // an array of promises when resolved becomes an array of numbers. where each number represent the count of all the approved attendees for the event with the same index in filteredEvents
+    const EventsAttendeesCountPromises = []
+
+    filteredEvents.map(evnt => {
+      EventsAttendeesCountPromises.push(AttendanceRequest.count({ event: evnt.id }, { status: `approved` }));
+    });
+
+
+    const EventsApprovedAttendeesCount = await Promise.all(EventsAttendeesCountPromises);
+
+    filteredEvents.forEach((evnt, i) => {
+      evnt._doc.attendees.approvedCount = EventsApprovedAttendeesCount[i];
+    });
+
+    res.status(200).json({ city, events: filteredEvents });
   } catch (err) {
     next(err);
   }
 });
 
+
+// ==========================================================
+// users are authenticated but access is not restricted
+// ==========================================================
+router.use(require("../middleware/auth.middleware"));
+// ==========================================================
+
 // get event by id
-router.get(`/:eventId`, validateId, async (req, res, next) => {
+router.get(`/:eventId`, validateIds, async (req, res, next) => {
   try {
+    const { user } = req;
     const { eventId } = req.params;
 
     const foundEvent = await Event.findById(eventId)
-    .populate({ path: `creator`, select: { password: 0, email: 0, __v: 0 } });
+      .populate({ path: `creator`, select: { password: 0, email: 0, __v: 0 } });
 
     if (!foundEvent) {
       handleNotExist(`event`, eventId, res);
       return;
     }
 
+    if (foundEvent.creator.id === user?.id) {
+      // appends all attendance requests to the found event
+      foundEvent._doc.attendees.requests = await AttendanceRequest.find({ event: eventId });
+    } else {
+      // appends all approved attendance requests to the found event
+      foundEvent._doc.attendees.approved = await AttendanceRequest.find({ event: eventId }, { status: `approved` });
+    }
+
```

Comment:
I need explanation for that haha

Reply:
I used ._doc because mongoose uses getters so as to allow document to be modified directly.

So the actual document is stored in: foundEvent._doc

Reply:
If mongoose didn't use getters we would be able to do it by doing:
foundEvent.attendees.approved = <approved requests>